pkg/group/manager/ldap/ldap.go

```diff
@@ -0,0 +1,396 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package ldap
+
+import (
+	"bytes"
+	"context"
+	"crypto/tls"
+	"fmt"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/Masterminds/sprig"
+	grouppb "github.com/cs3org/go-cs3apis/cs3/identity/group/v1beta1"
+	userpb "github.com/cs3org/go-cs3apis/cs3/identity/user/v1beta1"
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/errtypes"
+	"github.com/cs3org/reva/pkg/group"
+	"github.com/cs3org/reva/pkg/group/manager/registry"
+	"github.com/go-ldap/ldap/v3"
+	"github.com/mitchellh/mapstructure"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	registry.Register("ldap", New)
+}
+
+type manager struct {
+	c            *config
+	groupfilter  *template.Template
+	memberfilter *template.Template
+}
+
+type config struct {
+	Hostname        string     `mapstructure:"hostname"`
+	Port            int        `mapstructure:"port"`
+	BaseDN          string     `mapstructure:"base_dn"`
+	GroupFilter     string     `mapstructure:"groupfilter"`
+	MemberFilter    string     `mapstructure:"memberfilter"`
+	AttributeFilter string     `mapstructure:"attributefilter"`
+	FindFilter      string     `mapstructure:"findfilter"`
+	BindUsername    string     `mapstructure:"bind_username"`
+	BindPassword    string     `mapstructure:"bind_password"`
+	Idp             string     `mapstructure:"idp"`
+	Schema          attributes `mapstructure:"schema"`
+}
+
+type attributes struct {
+	// DN is the distinguished name in ldap, e.g. `cn=admins,ou=groups,dc=example,dc=org`
+	DN string `mapstructure:"dn"`
+	// UID is an immutable group id, see https://docs.microsoft.com/en-us/azure/active-directory/hybrid/plan-connect-design-concepts
+	GID string `mapstructure:"gid"`
+	// CN is the group name, typically `cn`, `gid` or `samaccountname`
+	CN string `mapstructure:"cn"`
+	// Mail is the email address of a group
+	Mail string `mapstructure:"mail"`
+	// Displayname is the Human readable name, e.g. `Database Admins`
+	DisplayName string `mapstructure:"displayName"`
+	// GIDNumber is a numeric id that maps to a filesystem gid, eg. 654321
+	GIDNumber string `mapstructure:"gidNumber"`
+}
+
+// Default attributes (Active Directory)
+var ldapDefaults = attributes{
+	DN:          "dn",
+	GID:         "objectGUID", // you can fall back to objectguid or even samaccountname but you will run into trouble when user names change. You have been warned.
```

Comment:
This GID should be an immutable, non reassignable, unique and persistent identifier for a group. The `ms-DS-ConsistencyGuid` we use for a users `UID` is made for that purpose. In general the defaults used in the ldap user and group provider should follow the same schema: either be Active Directory compatible or just use posixaccount and posixgroup. The latter schemas don't really have a unique attribute, which is why we recommend using a dedicated `ownclouduuid` attribute: https://github.com/owncloud/ocis/blob/c8668e8cb171860c70fec29e5ae945bca44f1fb7/deployments/examples/cs3_users_ocis/config/ldap/ldif/10_owncloud_schema.ldif

Reply:
I got this from https://docs.microsoft.com/en-us/azure/active-directory/hybrid/plan-connect-design-concepts#using-ms-ds-consistencyguid-as-sourceanchor.
> Use ms-DS-ConsistencyGuid as the sourceAnchor attribute for User objects. ObjectGUID is used for other object types.

But I can change it back to `ms-DS-ConsistencyGuid`.